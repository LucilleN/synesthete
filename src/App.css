.App {
  /* margin-left: 5rem;
  margin-right: 5rem; */
  width: 100%;
  /* padding-left: 5rem;
  padding-right: 5rem; */
}
