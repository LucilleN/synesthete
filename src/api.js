let api = 'https://misconfigured-app.com/'
// const API_KEY = 'dc6zaTOxFJmzC' // Giphy's public beta key (thank you Giphy).
const authorizationUrl = 'https://accounts.spotify.com/api/token'
const authorizationKey = 'Basic MTU5YWM4OGIxYzUzNGVkN2FlNDE2MDJmMWU1NThhNDk6YmUxOTFmMTg3ZGZlNDgzMjg3MDAxZDNhNWZlYTEyNTM='
const authorizationHeaders = {
  'Content-Type': 'application/x-www-form-urlencoded',
  'Authorization': authorizationKey
}

const getAccessToken = () => {
  const response = fetch(authorizationUrl, {
    authorizationHeaders
  }).then(okCheck, emitNativeError)
    .then(response => response.json())

  return response.access_token
}

const apiHost = host => { api = host }
const urlFor = resource => `${api}${resource}`

const HTTP_OK = 200

const throwResponseError = response => {
  const error = new Error(response.statusText)
  error.response = response
  throw error
}

const emitNativeError = error => {
  throw error
}

const statusCheck = successStatuses => response => {
  if (successStatuses.includes(response.status)) {
    return response
  } else {
    throwResponseError(response)
  }
}

const okCheck = statusCheck([HTTP_OK])

const getHeaders = () => {
  const token = getAccessToken()
  const headers = {
    'Authorization': `Bearer ${token}`
  }
  return headers
}

//
// const paramsWithApiKey = params => {
//   const result = new URLSearchParams(params)
//   result.set('api_key', API_KEY)
//   return result
// }

// The fetch function initiates a connection to the web service.
// fetch returns a _promise_: an object that represents a future result.
// Thus, the function actually returns right away. However, when the
// anticipated result does show up, the code specifies what to do using
// either `then` or `catch`. Both functions accept another function,
// to be called upon a successful or failed promise, respectively.
// Furthermore, then `then` function can be chained: its return result
// is passed to the next `then` function as an argument. Here, the initial
// handler converts the raw result into JSON. That JSON then goes to the
// next handler, which does the actual work of putting the result on the
// web page.
//
// The design of fetch allows this entire sequence to be rendered in a
// _single statement_, thus obviating the need for curly braces but
// resulting in what many will view to be a decrease in readability
// (for those who arenâ€™t used to functional-style programming). YMMV

// const query = (resource, params) => fetch(`${urlFor(resource)}?${paramsWithApiKey(params)}`, {
const query = (resource, params) => {
  const optionalParams = (params) ? `?${new URLSearchParams(params)}` : ""

  const headers = getHeaders()

  fetch(`${urlFor(resource)}${optionalParams}`, {
    headers
  }).then(okCheck, emitNativeError)
    .then(response => response.json())
}

// const searchGifs = params => query('gifs/search', params)
const searchSongs = params => query('search', params)

const getRecommendation = params => query('recommendations', params)

const getAudioFeatures = songID => query(`audio-features/${songID}`)


export {
  apiHost,
  // searchGifs
  searchSongs,
  getAudioFeatures,
  getRecommendation
}
