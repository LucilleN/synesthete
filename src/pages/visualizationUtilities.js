export const loadMusic = ({
  audioRef,
  canvasRef,
  fileRef,
  srcUrl,
  existingContext,
  setContext,
  existingAudioNode,
  setAudioNode,
  existingAnalyser,
  setAnalyser,
  setError,
  keyOffset=0
}) => {

  const audio = audioRef.current
  if (!audio) {
    return
  }

  if (fileRef) {
    try {
      const file = fileRef.current
      if (file && file.files.length > 0) {
        audio.src = URL.createObjectURL(file.files[0])
      }
    } catch (error) {
      setError("Sorry, but something went wrong.")
    }
  }
  else {
    audio.src = srcUrl
  }

  const canvas = canvasRef.current
  canvas.width = window.innerWidth
  canvas.height = window.innerHeight
  const ctx = canvas.getContext("2d")

  const context = (existingContext) ? existingContext : new AudioContext()
  const audioNode = (existingAudioNode) ? existingAudioNode : context.createMediaElementSource(audio)
  const analyser = (existingAnalyser) ? existingAnalyser : context.createAnalyser()

  audioNode.connect(analyser)

  analyser.connect(context.destination) // End destination of an audio graph in a given context
  analyser.fftSize = 16384

  const bufferLength = analyser.frequencyBinCount
  const dataArray = new Uint8Array(bufferLength)

  const colors = [
    [179, 0, 54], // red
    [255, 60, 0], // red-orange
    [255, 123, 0], // orange
    [255, 178, 0], // honey yellow
    [255, 243, 163], // light yellow
    [203, 247, 57], // yellow-green
    [25, 191, 0], // green
    [0, 161, 155], // blue-green
    [0, 181, 169], // green-blue
    [0, 113, 212], // blue
    [39, 16, 173], // indigo
    [75, 0, 145], // dark purple
    [145, 10, 145], // magenta
  ]

  function getColorOfSoundBars(keyIndex, offset) {
    let newIndex = keyIndex + offset
    if (newIndex >= colors.length) {
      newIndex -= colors.length
    }
    if (newIndex < 0) {
      newIndex += colors.length
    }
    return colors[newIndex]
  }

  const canvasWidth = canvas.width
  const canvasHeight = canvas.height
  const totalNumberOfBars = 120
  const barWidth = (canvasWidth / bufferLength) * 40
  const barHeightMax = 500
  const barSpacing = 10
  const barsPerColorSection = 6

  let barHeight
  let x = 0

  function renderFrame() {
    requestAnimationFrame(renderFrame)

    x = 0

    analyser.getByteFrequencyData(dataArray)

    // Clears canvas with black and use opacity 0.1 to create fade effect
    ctx.fillStyle = "rgba(0,0,0,0.1)"
    ctx.fillRect(0, 0, canvasWidth, canvasHeight)

    for (let barIndex = 0; barIndex < totalNumberOfBars; barIndex++) {
      barHeight = (dataArray[barIndex] * 2)

      // let offset = -10 // Lord Huron, the Night We Met (key: A major)
      // let offset = -7 // Kina Grannis, Iris (key: F# Major)

      // let [r, g, b] = getColorOfSoundBars(Math.floor(barIndex/barsPerColorSection), offset)
      let [r, g, b] = getColorOfSoundBars(Math.floor(barIndex/barsPerColorSection), keyOffset)
      let a = (barHeight / barHeightMax) ** 3.5

      ctx.fillStyle = `rgba(${r},${g},${b},${a})`
      ctx.fillRect(x, (canvasHeight - barHeight), barWidth, barHeight)

      x += barWidth + barSpacing
    }
  }

  audio.play()
  renderFrame()

  if (!existingAudioNode) {
    setAudioNode(audioNode)
  }
  if (!existingAnalyser) {
    setAnalyser(analyser)
  }
  if (!existingContext) {
    setContext(context)
  }
}
