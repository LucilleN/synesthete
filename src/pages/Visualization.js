import React, { useState, useEffect, useRef } from 'react'
import Typography from '@material-ui/core/Typography'
import { withStyles } from '@material-ui/core/styles'
import RecommendationButton from '../components/RecommendationButton'
import StartButton from '../components/StartButton'
import ErrorDialog from '../components/ErrorDialog'

import { Redirect } from 'react-router-dom'

import { getRecommendation } from '../api'
import { loadMusic, performAudioFeaturesQuery } from './visualizationUtilities'
import { defaultErrorText } from '../App'

export const styles = theme => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center'
  },
  fileInput: {
    zIndex: 3,
    color: 'white'
  },
  urlInput: {
    zIndex: 3,
    color: theme.palette.dark.pink,
    width: 250,
    margin: 15
  },
  urlButton: {
    zIndex: 3
  },
  recommendationButton: {
    zIndex: 10,
    marginTop: 200
  },
  canvas: {
    position: 'fixed',
    top: '60px',
    left: '0px',
    width: '100%',
    height: 'calc(100%-60px)',
    background: 'black',
    zIndex: 0
  },
  audio:{
    position: 'fixed',
    bottom: '20px',
    width: '90vw',
    height: '35px',
    zIndex: 3,
    opacity: 0.1,
    '&:hover': {
       opacity: 0.8,
    }
  },
  titleBar: {
    width: '100%',
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  title: {
    zIndex: 4,
    color: theme.palette.light.pink,
    fontSize: '3rem',
    textAlign: 'center',
    marginTop: 5
  },
  subtitle: {
    zIndex: 4,
    color: theme.palette.dark.pink,
    fontSize: '1.75rem',
    textAlign: 'center',
    marginBottom: 8,
    marginTop: -32
  },
  songInfoContainer: {
    zIndex: 4,
    width: '100%'
  },
  songInfo: {
    zIndex: 4,
    color: theme.palette.light.pink,
    fontSize: '0.75rem',
    textAlign: 'left',
    width: '100%',
    marginLeft: 8
  },
  loadOptionButton: {
    zIndex: 4,
    color: theme.palette.white,
    fontSize: '1rem',
    background: theme.palette.dark.pink,
    height: '60px',
    width: '250px',
    borderRadius: '30px',
    border: 'none',
    margin: '10px 20px',
    '&:hover': {
       background: theme.palette.white,
       color: theme.palette.dark.pink
    }
  },
  buttonBar: {
    width: '100%',
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-evenly'
  },
  button: {
    zIndex: 4,
    color: theme.palette.white,
    fontSize: '1rem',
    background: theme.palette.dark.purple,
    height: '60px',
    width: '250px',
    borderRadius: '30px',
    border: 'none',
    margin: '10px 20px',
    textTransform: 'none',
    '&:hover': {
       background: theme.palette.white,
       color: theme.palette.dark.purple
    }
  }
})

const Visualization = props => {
  const { classes } = props
  const { trackObject } = props.location.state

  if (!trackObject) {
    console.log("no track object")
    setError(defaultErrorText)
  }

  // May need this in the future to allow users to access a song's visualization just by
  // clicking on a saved link rather than going through the search every time, but this
  // would require another API call to get the trackObject based on the songID, so I
  // chose not to implement it for this project.
  // const { songID } = props.match.params

  const audioRef = useRef(null)
  const canvasRef = useRef(null)

  const [error, setError] = useState(null)
  const [audioFeatures, setAudioFeatures] = useState(null)
  const [recommendedSong, setRecommendedSong] = useState(null)

  const [existingAudioNode, setAudioNode] = useState(null)
  const [existingAnalyser, setAnalyser] = useState(null)
  const [existingContext, setContext] = useState(null)

  const [canLoadMusic, setCanLoadMusic] = useState(false)

  useEffect(() => {
    if (trackObject) {
      performAudioFeaturesQuery({
        setError,
        songID: trackObject.id,
        setAudioFeatures,
        defaultErrorText
      })
    }
  }, [trackObject])

  useEffect(() => {
    setCanLoadMusic(true)
  }, [audioFeatures])

  useEffect(() => {
    if (canLoadMusic && audioFeatures) {
      loadMusic({
        audioRef,
        canvasRef,
        srcUrl: trackObject.preview_url,
        existingContext,
        setContext,
        existingAudioNode,
        setAudioNode,
        existingAnalyser,
        setAnalyser,
        setError,
        audioFeatures
      })
    }
    setCanLoadMusic(false)
  }, [canLoadMusic])

  const performRecommendationQuery = async event => {
    event.preventDefault()
    setError(null)

    let currentRecommendation

    try {
      const result = await getRecommendation({
        limit: 1,
        seed_tracks: `${trackObject.id}`,
        seed_artists: `${trackObject.artists[0].id}`,
        target_energy: `${audioFeatures.energy}`,
        target_valence: `${audioFeatures.valence}`,
        target_acousticness: `${audioFeatures.acousticness}`,
        target_key: `${audioFeatures.key}`
      })
      currentRecommendation = result.tracks[0]

      const isSameSong = (track1, track2) => {
        return (
          track1.id === track2.id &&
          track1.artists[0].id === track2.artists[0].id &&
          track1.title === track2.title
        )
      }

      // Recommendation might be a song with a null preview_url; if so, recommend another one
      // Recommendation might also be the current song itself; if so, recommend another
      while (!currentRecommendation.preview_url || isSameSong(trackObject, currentRecommendation)) {
        console.log("currentTrack's id", trackObject.id)
        console.log("currentTrack's preview_url", trackObject.preview_url)
        console.log("currentRecommendation.id", currentRecommendation.id)
        const result = await getRecommendation({
          limit: 3,
          seed_tracks: `${currentRecommendation.id}`,
          // seed_artists: `${currentRecommendation.artists[0].id}`
        })
        currentRecommendation = result.tracks[0]
        console.log("got a new recommendation with id", currentRecommendation.id)
        console.log("new recommendation's preview url", currentRecommendation.preview_url)
      }

      setRecommendedSong(result.tracks[0])

    } catch (error) {
      setError(defaultErrorText)
    }
  }

  return (
    <div className={classes.root}>
      <canvas ref={canvasRef} id="canvas" width="300" height="300" className={classes.canvas}></canvas>
      <audio ref={audioRef} id="audio" crossOrigin="anonymous" controls className={classes.audio}></audio>
      <div className={classes.titleBar}>
        <StartButton href="/search" text="< new search"/>
        <Typography className={classes.title}>
          Visualization:<br></br>
        </Typography>
        <RecommendationButton className={classes.recommendationButton} handleClick={performRecommendationQuery}/>
      </div>
      <Typography className={classes.subtitle}>
        "{trackObject.name}" by {trackObject.artists[0].name}
      </Typography>
      {recommendedSong && (
        <Redirect to={{
            pathname: `/visualization/${recommendedSong.id}`,
            state: { trackObject: recommendedSong }
          }}
        />
      )}
      {error && (
        <div id="error">
          <ErrorDialog error={error} errorSubtitle={"The Visualizer doesn't work when it's been loaded externally. Try refreshing the page or starting a new search."}/>
        </div>
      )}
    </div>
  )
}

export default withStyles(styles)(Visualization)
