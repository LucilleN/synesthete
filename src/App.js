import React, { useEffect } from 'react'
import { Route, BrowserRouter as Router, Switch } from 'react-router-dom'

import { MuiThemeProvider, createMuiTheme } from '@material-ui/core/styles'
import { withStyles } from '@material-ui/core/styles'

import Homepage from './pages/Homepage'
import About from './pages/About'
import Search from './pages/Search'
import UploadVisualization from './pages/UploadVisualization'
import Visualization from './pages/Visualization'
import MenuBar from './components/MenuBar'

import { apiHost } from './api'

export const theme = createMuiTheme({
  palette: {
    primary: {
      main: '#c23063', // magenta
      dark: '#180142' // dark purple
    },
    type: 'dark',
    dark: {
      pink: '#bf265c',
      purple: '#1c004f',
    },
    medium: {
      purple: '#5a0b66',
    },
    light: {
      pink: '#db91aa',
      purple: '#ac91db'
    },
    white: '#FFFFFF',
  },
  typography: {
    fontFamily: ['Dosis', 'Roboto', 'Open Sans', '"Helvetica Neue"', 'Arial', 'sans-serif'].join(',')
  }
})

export const defaultErrorText = 'Sorry, but something went wrong.'

const styles = () => ({
  root: {
    width: '100vw',
    height: '100%',
    minHeight: '100vh',
    background: `linear-gradient(0deg, ${theme.palette.dark.pink} 0%, ${theme.palette.dark.purple} 80%)`
  }
})

const routes = (
  <Router>
    <div>
      <MenuBar />
      <Switch>
        <Route exact path="/" component={Homepage} />
        <Route exact path="/about" component={About} />
        <Route exact path="/search" component={Search} />
        <Route exact path="/upload" component={UploadVisualization} />
        <Route exact path="/visualization" component={Visualization} />
        <Route exact path="/visualization/:songID" component={Visualization} />
      </Switch>
    </div>
  </Router>
)

const App = (props) => {

  useEffect(() => apiHost('https://api.spotify.com/v1/'))

  const { classes } = props

  return (
    <div className={classes.root}>
      <MuiThemeProvider theme={theme}>
        {routes}
      </MuiThemeProvider>
    </div>
  )
}

export default withStyles(styles)(App)
