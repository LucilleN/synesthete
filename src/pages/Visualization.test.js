import React from 'react'
import ReactDOM from 'react-dom'
import TestRenderer from 'react-test-renderer'
import ReactTestUtils from 'react-dom/test-utils'
import { MuiThemeProvider } from '@material-ui/core/styles'

import { Route, BrowserRouter as Router } from 'react-router-dom'

import sinon from 'sinon'

import { theme } from '../App'
import Visualization from './Visualization'
import * as visualizationUtilities from './visualizationUtilities'
import * as api from '../api'

const track = {
  "album" : {
    "album_type" : "album",
    "artists" : [ {
      "external_urls" : {
        "spotify" : "https://open.spotify.com/artist/6ltzsmQQbmdoHHbLZ4ZN25"
      },
      "href" : "https://api.spotify.com/v1/artists/6ltzsmQQbmdoHHbLZ4ZN25",
      "id" : "6ltzsmQQbmdoHHbLZ4ZN25",
      "name" : "Lord Huron",
      "type" : "artist",
      "uri" : "spotify:artist:6ltzsmQQbmdoHHbLZ4ZN25"
    } ],
    "available_markets" : [ "CA", "US" ],
    "external_urls" : {
      "spotify" : "https://open.spotify.com/album/3yoNZlqerJnsnMN5EDwwBS"
    },
    "href" : "https://api.spotify.com/v1/albums/3yoNZlqerJnsnMN5EDwwBS",
    "id" : "3yoNZlqerJnsnMN5EDwwBS",
    "images" : [ {
      "height" : 640,
      "url" : "https://i.scdn.co/image/d0a15f9288814a7e3f199ab2d1a6b0c9a9a633b5",
      "width" : 640
    }, {
      "height" : 300,
      "url" : "https://i.scdn.co/image/9bcea4ada214a061cecce6d565319f839ea83ce1",
      "width" : 300
    }, {
      "height" : 64,
      "url" : "https://i.scdn.co/image/0a2f11966d40cd6fc87ee70b71d8547578e87fe8",
      "width" : 64
    } ],
    "name" : "Strange Trails",
    "release_date" : "2015-04-07",
    "release_date_precision" : "day",
    "total_tracks" : 14,
    "type" : "album",
    "uri" : "spotify:album:3yoNZlqerJnsnMN5EDwwBS"
  },
  "artists" : [ {
    "external_urls" : {
      "spotify" : "https://open.spotify.com/artist/6ltzsmQQbmdoHHbLZ4ZN25"
    },
    "href" : "https://api.spotify.com/v1/artists/6ltzsmQQbmdoHHbLZ4ZN25",
    "id" : "6ltzsmQQbmdoHHbLZ4ZN25",
    "name" : "Lord Huron",
    "type" : "artist",
    "uri" : "spotify:artist:6ltzsmQQbmdoHHbLZ4ZN25"
  } ],
  "available_markets" : [ "CA", "US" ],
  "disc_number" : 1,
  "duration_ms" : 208211,
  "explicit" : false,
  "external_ids" : {
    "isrc" : "US53Q1200103"
  },
  "external_urls" : {
    "spotify" : "https://open.spotify.com/track/0QZ5yyl6B6utIWkxeBDxQN"
  },
  "href" : "https://api.spotify.com/v1/tracks/0QZ5yyl6B6utIWkxeBDxQN",
  "id" : "0QZ5yyl6B6utIWkxeBDxQN",
  "is_local" : false,
  "name" : "The Night We Met",
  "popularity" : 76,
  "preview_url" : "https://p.scdn.co/mp3-preview/1c0da00b5c95a1a6c9dfc05b14a1a628a6e0ad73?cid=159ac88b1c534ed7ae41602f1e558a49",
  "track_number" : 14,
  "type" : "track",
  "uri" : "spotify:track:0QZ5yyl6B6utIWkxeBDxQN"
}

const audioFeatures = {
  "danceability": 0.441,
  "energy": 0.379,
  "key": 2,
  "loudness": -9.545,
  "mode": 1,
  "speechiness": 0.0449,
  "acousticness": 0.968,
  "instrumentalness": 0.262,
  "liveness": 0.639,
  "valence": 0.105,
  "tempo": 174.118,
  "type": "audio_features",
  "id": "0QZ5yyl6B6utIWkxeBDxQN",
  "uri": "spotify:track:0QZ5yyl6B6utIWkxeBDxQN",
  "track_href": "https://api.spotify.com/v1/tracks/0QZ5yyl6B6utIWkxeBDxQN",
  "analysis_url": "https://api.spotify.com/v1/audio-analysis/0QZ5yyl6B6utIWkxeBDxQN",
  "duration_ms": 208227,
  "time_signature": 3
}

const recommendation = {
  "tracks" : [ {
    "artists" : [ {
      "external_urls" : {
        "spotify" : "https://open.spotify.com/artist/02YLJJnWC7YQVixkjEBRn7"
      },
      "href" : "https://api.spotify.com/v1/artists/02YLJJnWC7YQVixkjEBRn7",
      "id" : "02YLJJnWC7YQVixkjEBRn7",
      "name" : "Billie Marten",
      "type" : "artist",
      "uri" : "spotify:artist:02YLJJnWC7YQVixkjEBRn7"
    } ],
    "available_markets" : [ "AD", "AE", "AR", "AT", "AU", "BE", "BG", "BH", "BO", "BR", "CA", "CH", "CL", "CO", "CR", "CY", "CZ", "DE", "DK", "DO", "DZ", "EC", "EE", "EG", "ES", "FI", "FR", "GB", "GR", "GT", "HK", "HN", "HU", "ID", "IE", "IL", "IN", "IS", "IT", "JO", "JP", "KW", "LB", "LI", "LT", "LU", "LV", "MA", "MC", "MT", "MX", "MY", "NI", "NL", "NO", "NZ", "OM", "PA", "PE", "PH", "PL", "PS", "PT", "PY", "QA", "RO", "SA", "SE", "SG", "SK", "SV", "TH", "TN", "TR", "TW", "US", "UY", "VN", "ZA" ],
    "disc_number" : 1,
    "duration_ms" : 187266,
    "explicit" : false,
    "external_ids" : {
      "isrc" : "GBARL1600328"
    },
    "external_urls" : {
      "spotify" : "https://open.spotify.com/track/0DwaOWI1kC2s98qLj0LTDd"
    },
    "href" : "https://api.spotify.com/v1/tracks/0DwaOWI1kC2s98qLj0LTDd",
    "id" : "0DwaOWI1kC2s98qLj0LTDd",
    "is_local" : false,
    "name" : "La Lune",
    "popularity" : 60,
    "preview_url" : "https://p.scdn.co/mp3-preview/14399ace213706a367f6bd9f34357dd5f532a588?cid=159ac88b1c534ed7ae41602f1e558a49",
    "track_number" : 1,
    "type" : "track",
    "uri" : "spotify:track:0DwaOWI1kC2s98qLj0LTDd"
  } ],
  "seeds" : [ {
    "initialPoolSize" : 250,
    "afterFilteringSize" : 250,
    "afterRelinkingSize" : 250,
    "id" : "6ltzsmQQbmdoHHbLZ4ZN25",
    "type" : "ARTIST",
    "href" : "https://api.spotify.com/v1/artists/6ltzsmQQbmdoHHbLZ4ZN25"
  }, {
    "initialPoolSize" : 249,
    "afterFilteringSize" : 249,
    "afterRelinkingSize" : 249,
    "id" : "0QZ5yyl6B6utIWkxeBDxQN",
    "type" : "TRACK",
    "href" : "https://api.spotify.com/v1/tracks/0QZ5yyl6B6utIWkxeBDxQN"
  } ]
}

let component
beforeEach(async () => {
  await TestRenderer.act(async () => {
    component = TestRenderer.create(
      <MuiThemeProvider theme={theme}>
        <Visualization location={{ state: { trackObject: track } }}/>
      </MuiThemeProvider>
    )
  })
})

describe('initial state', () => {
  it('should have a button to navigate back to search and a button to get a recommendation', () => {
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should display the title and artist of the current song being played', () => {
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render a canvas', () => {
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render an audio element', () => {
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('loading music', () => {
  let div
  let visualizationUtilitiesStub
  beforeEach(async () => {
    sinon.stub(visualizationUtilities, 'loadMusic')
    sinon.stub(api, 'getAudioFeatures')

    api.getAudioFeatures.returns(Promise.resolve(audioFeatures))

    div = document.createElement('div')
    await ReactTestUtils.act(async () => {
      await ReactDOM.render(
        <MuiThemeProvider theme={theme}>
          <Visualization location={{ state: { trackObject: track } }}/>
        </MuiThemeProvider>
      , div)
    })
  })

  afterEach(() => {
    ReactDOM.unmountComponentAtNode(div)
    visualizationUtilities.loadMusic.restore()
    api.getAudioFeatures.restore()
  })

  it('still renders a canvas', () => {
    const canvas = div.querySelector('#canvas')
    expect(canvas).not.toBeNull()
  })

  it('gets the audio features', () => {
    expect(api.getAudioFeatures.called).toBe(true)
  })

  it('automatically loads music', (done) => {
    let counter = 0
    const nextEffect = () => {
      if (counter < 4) {
        counter += 1
        process.nextTick(nextEffect)
      } else {
        expect(visualizationUtilities.loadMusic.called).toBe(true)
        done()
      }
    }

    process.nextTick(nextEffect)
  })
})

describe('the recommendation button', () => {
  let div
  beforeEach(async () => {
    sinon.stub(visualizationUtilities, 'loadMusic')
    sinon.stub(api, 'getRecommendation')

    api.getRecommendation.returns(Promise.resolve(recommendation))

    div = document.createElement('div')
    await ReactTestUtils.act(async () => {
      await ReactDOM.render(
        <Router>
          <MuiThemeProvider theme={theme}>
            <Visualization location={{ state: { trackObject: track } }}/>
          </MuiThemeProvider>
        </Router>
      , div)
    })
    const recommendationButton = div.querySelector("#recommendation-button")
    await ReactTestUtils.act(async() => {
      await ReactTestUtils.Simulate.click(recommendationButton)
    })

  })

  afterEach(() => {
    ReactDOM.unmountComponentAtNode(div)
    visualizationUtilities.loadMusic.restore()
    api.getRecommendation.restore()
  })

  it('should call the getRecommendation function when clicked', () => {
    let counter = 0
    const nextEffect = () => {
      if (counter < 2) {
        counter += 1
        process.nextTick(nextEffect)
      } else {
        expect(api.getRecommendation.called).toBe(true)
        done()
      }
    }

    process.nextTick(nextEffect)


  })
})
