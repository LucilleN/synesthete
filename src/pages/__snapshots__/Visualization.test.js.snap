// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial state should display the title and artist of the current song being played 1`] = `null`;

exports[`initial state should have a button to navigate back to search and a button to get a recommendation 1`] = `
<div
  className="Visualization-root-1"
>
  <canvas
    className="Visualization-canvas-6"
    height="300"
    id="canvas"
    width="300"
  />
  <audio
    className="Visualization-audio-7"
    controls={true}
    crossOrigin="anonymous"
    id="audio"
  />
  <div
    className="Visualization-titleBar-8"
  >
    <a
      aria-disabled={false}
      aria-label="Start"
      className="MuiButtonBase-root MuiFab-root StartButton-fab-16 MuiFab-extended MuiFab-primary"
      href="/search"
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
    >
      <span
        className="MuiFab-label"
      >
         
        <p
          className="MuiTypography-root StartButton-text-17 MuiTypography-body1"
        >
          &lt; new search
        </p>
      </span>
    </a>
    <p
      className="MuiTypography-root Visualization-title-9 MuiTypography-body1"
    >
      Visualization:
      <br />
    </p>
    <a
      aria-disabled={false}
      aria-label="Get Recommendation"
      className="MuiButtonBase-root MuiFab-root RecommendationButton-fab-61 MuiFab-extended MuiFab-primary"
      href="/visualization"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
    >
      <span
        className="MuiFab-label"
      >
         
        <p
          className="MuiTypography-root RecommendationButton-text-62 MuiTypography-body1"
        >
          get a visually similar song
        </p>
      </span>
    </a>
  </div>
  <p
    className="MuiTypography-root Visualization-subtitle-10 MuiTypography-body1"
  >
    "
    The Night We Met
    " by 
    Lord Huron
  </p>
</div>
`;

exports[`initial state should render a canvas 1`] = `
<div
  className="Visualization-root-63"
>
  <canvas
    className="Visualization-canvas-68"
    height="300"
    id="canvas"
    width="300"
  />
  <audio
    className="Visualization-audio-69"
    controls={true}
    crossOrigin="anonymous"
    id="audio"
  />
  <div
    className="Visualization-titleBar-70"
  >
    <a
      aria-disabled={false}
      aria-label="Start"
      className="MuiButtonBase-root MuiFab-root StartButton-fab-78 MuiFab-extended MuiFab-primary"
      href="/search"
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
    >
      <span
        className="MuiFab-label"
      >
         
        <p
          className="MuiTypography-root StartButton-text-79 MuiTypography-body1"
        >
          &lt; new search
        </p>
      </span>
    </a>
    <p
      className="MuiTypography-root Visualization-title-71 MuiTypography-body1"
    >
      Visualization:
      <br />
    </p>
    <a
      aria-disabled={false}
      aria-label="Get Recommendation"
      className="MuiButtonBase-root MuiFab-root RecommendationButton-fab-123 MuiFab-extended MuiFab-primary"
      href="/visualization"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
    >
      <span
        className="MuiFab-label"
      >
         
        <p
          className="MuiTypography-root RecommendationButton-text-124 MuiTypography-body1"
        >
          get a visually similar song
        </p>
      </span>
    </a>
  </div>
  <p
    className="MuiTypography-root Visualization-subtitle-72 MuiTypography-body1"
  >
    "
    The Night We Met
    " by 
    Lord Huron
  </p>
</div>
`;

exports[`initial state should render an audio element 1`] = `
<div
  className="Visualization-root-1"
>
  <canvas
    className="Visualization-canvas-6"
    height="300"
    id="canvas"
    width="300"
  />
  <audio
    className="Visualization-audio-7"
    controls={true}
    crossOrigin="anonymous"
    id="audio"
  />
  <div
    className="Visualization-titleBar-8"
  >
    <a
      aria-disabled={false}
      aria-label="Start"
      className="MuiButtonBase-root MuiFab-root StartButton-fab-16 MuiFab-extended MuiFab-primary"
      href="/search"
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
    >
      <span
        className="MuiFab-label"
      >
         
        <p
          className="MuiTypography-root StartButton-text-17 MuiTypography-body1"
        >
          &lt; new search
        </p>
      </span>
    </a>
    <p
      className="MuiTypography-root Visualization-title-9 MuiTypography-body1"
    >
      Visualization:
      <br />
    </p>
    <a
      aria-disabled={false}
      aria-label="Get Recommendation"
      className="MuiButtonBase-root MuiFab-root RecommendationButton-fab-61 MuiFab-extended MuiFab-primary"
      href="/visualization"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
    >
      <span
        className="MuiFab-label"
      >
         
        <p
          className="MuiTypography-root RecommendationButton-text-62 MuiTypography-body1"
        >
          get a visually similar song
        </p>
      </span>
    </a>
  </div>
  <p
    className="MuiTypography-root Visualization-subtitle-10 MuiTypography-body1"
  >
    "
    The Night We Met
    " by 
    Lord Huron
  </p>
</div>
`;

exports[`loading music should automatically load music 1`] = `null`;
